Add the notion of @_implementationOnly overrides

jrose-apple opened the pull request:
When an `@_implementationOnly` import includes Objective-C categories for existing types, it's useful to be able to override the members provided in those categories without exposing them to clients of the framework being built. Allow this as long as the overriding declaration is marked as `@_implementationOnly` itself, with an additional check that the type of the declaration does not change. (Normally overrides are allowed to change in covariant ways.)

This is definitely not sound when not using `-enable-library-evolution`, and when it is using `-enable-library-evolution` it's probably relying on the fact that overrides of ObjC things are always `dynamic` and that overridable Swift members can never be added in extensions. But the feature is still underscored and there's a definite need for it.

Part of rdar://50827914

jrose-apple committed:
Add the notion of @_implementationOnly overrides

When an @_implementationOnly import includes Objective-C categories
for existing types, it's useful to be able to override the members
provided in those categories without exposing them to clients of the
framework being built. Allow this as long as the overriding
declaration is marked as @_implementationOnly itself, with an
additional check that the type of the declaration does not change.
(Normally overrides are allowed to change in covariant ways.)

Part of rdar://50827914

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
Follow-up work: when using `@_implementationOnly`, skip checking the type of a declaration too.

slavapestov commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2639,6 +2640,41 @@ void AttributeChecker::visitFunctionBuilderAttr(FunctionBuilderAttr *attr) {
+  if (!derivedInterfaceTy->isEqual(overrideInterfaceTy)) {
+    TC.diagnose(VD, diag::implementation_only_override_changed_type,
+                overrideInterfaceTy);
+    TC.diagnose(overridden, diag::overridden_here);
```
We tend to prefer calling diagnose() on the decl instead, I’m service of eventually removing unnecessary dependencies on the TypeChecker instance

jrose-apple replied:
Oh, I forgot that Decls have a direct `diagnose` now. Will update.

jrose-apple commented:
@swift-ci Please test Linux

jrose-apple commented:
@swift-ci Please smoke test macOS

jrose-apple commented:
@swift-ci Please test Linux

jrose-apple commented:
Holding off on merging until I've played with this on one of the Apple frameworks using the feature, to make sure it's behaving as expected.

jrose-apple committed:
Add test cases for when a property/subscript is redeclared readwrite

And improve the comment around the FIXME for this, which is tricky to
implement.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2630,6 +2631,69 @@ void AttributeChecker::visitFunctionBuilderAttr(FunctionBuilderAttr *attr) {
+    overrideInterfaceTy =
+        FunctionType::get(overrideInterfaceFuncTy->getParams(),
+                          overrideInterfaceFuncTy->getResult());
+  }
```
I am *not happy* with this whole block. Before, this section of the function was four lines that were a bit verbose but basically said what they did. Now there's a bunch of extra type munging for no obvious reason.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test
